File: 1239_maximum_length_of_concat_string/max_length.js
/**
 * Time:        O(2**n)
 * Space:       O(n)
 * @param       {string[]} arr
 * @return      {number}
 */
const maxLength = arr => {
  const res = [0]
  dfs(arr, 0, "", res)
  return res[0]
}

const dfs = (arr, index, tmp, res) => {
  if (index == arr.length) {
    if (uniqChars(tmp) > res[0]) res[0] = tmp.length
    return
  }
  dfs(arr, index + 1, tmp, res)
  dfs(arr, index + 1, tmp + arr[index], res)
}

const uniqChars = str => {
  const s = new Set(str)
  if (str.length === s.size || 0) return str.length
  return -1
}

File: 1239_maximum_length_of_concat_string/max_length.py
from typing import List


def max_length(arr: List[str]) -> int:
    """
    Time: O(2**n)
    Space: O(n)
    """
    res = [0]
    dfs(arr, 0, "", res)
    return res[0]


def dfs(arr: List[str], index: int, tmp: str, res: List[int]):
    if index == len(arr):
        if unique_chars(tmp) > res[0]:
            res[0] = len(tmp)
        return
    # Simulate adding current string and not
    dfs(arr, index + 1, tmp, res)
    dfs(arr, index + 1, tmp + arr[index], res)


def unique_chars(s):
    size = len(s)
    if size > len(set(s)):
        return -1
    return size

# WAYYY FASTER


def max_length_again(arr: List[str]) -> int:
    res = [set()]
    for a in arr:
        if len(a) > len(set(a)):
            continue
        a = set(a)
        for r in res[:]:
            if a & r:
                continue
            res.append(a | r)
    return max(len(s) for s in res)
